### PR TITLE
Implement authorization with HuggingFace [version for the 2nd Bengali run]

### PR DESCRIPTION
**Pull request by borzunov**
In addition to #6, this PR implements the following:

1. Explicit whitelisting by moderators (without using `robot-bengali` to add a new collaborator automatically).
2. Retries with exponential delays for unexpected (i.e. not just invalid username/password) exceptions during authorization.
3. Using coordinator_ip/port returned by the HuggingFace API (both as a DHT initial peer and syslog host) instead of hardcoding them.
4. A better CLI:
     - If username/password are invalid, the CLI shows a user-friendly message, then shows the inputs again.
     - If an email is used instead of a username (a popular misconception), the CLI shows a user-friendly message, then shows the input again.

Based on the protocol from https://github.com/learning-at-home/hivemind/issues/253 implemented in https://github.com/learning-at-home/hivemind/pull/255.

### CODE REVIEW

File: colab_starter.ipynb [hunk 1]
```diff
@@ -36,7 +36,7 @@
         "colab_type": "text"
       },
       "source": [
-        "<a href=\"https://colab.research.google.com/github/mryab/collaborative-training/blob/main/colab_starter.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
+        "<a href=\"https://colab.research.google.com/github/mryab/collaborative-training/blob/auth-v2/colab_starter.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
```
Comment (justheuristic) [inline colab_starter.ipynb]: upon merge, please make sure you update this link to the correct branch

**Review by justheuristic — APPROVED** [review]

File: colab_starter.ipynb [hunk 2]
```diff
@@ -69,34 +69,35 @@
       "source": [
         "experiment_name = \"bengali_test_1_auth\"\n",
         "hivemind_version = \"bengali_test_1_auth\"\n",
-        "collaborative_training_version = \"auth\" \n",
-        "coordinator_ip = '18.191.120.93'\n",
-        "coordinator_port = 31337\n",
+        "collaborative_training_version = \"auth-v2\"\n",
```
Comment (borzunov) [inline colab_starter.ipynb]: TODO: Update branch before merge

File: huggingface_auth.py [hunk 3]
```diff
@@ -113,20 +134,27 @@ def _token_to_bytes(access_token: AccessToken) -> bytes:
...
+        if experiment_id is None:
+            experiment_id = input('HuggingFace experiment ID: ')
+
+        username = os.getenv('HF_USERNAME')
+        if username is None:
+            while True:
+                username = input('HuggingFace username: ')
+                if '@' not in username:
+                    break
+                print('Please enter your Huggingface _username_ instead of the email address!')
```
Comment (SaulLu) [inline huggingface_auth.py:149]: Really great addition :+1:

**Review by SaulLu — APPROVED** [review]
Thank you very much for making all these changes so quickly! 

I don't have anything to say except about the password storage issue, but as discussed on the [other PR](https://github.com/mryab/collaborative-training/pull/6), I think it will be too short to find a solution/agree before tomorrow.

**Review by mryab — APPROVED** [review]

**Pull request merged**